pyuvdata/uvdata/mir_parser.py

```diff
@@ -10,1082 +10,6383 @@
+                continue
+
+            # If we are _not_ skipping the spectral averaging, then it turns out to
+            # be faster to convert the raw data to "regular" data, spectrally average
+            # it, and then convert it back to the raw format. Note that we set up a
+            # "dummy" dict here with an sphid of 0 to make it easy to retrieve that
+            # entry after the sequence of calls.
+            if return_vis:
+                data_dict[sphid] = MirParser._rechunk_data(
+                    MirParser.convert_raw_to_vis({0: sp_raw}),
+                    [chan_avg],
+                    inplace=True,
+                )[0]
+            else:
+                data_dict[sphid] = MirParser.convert_vis_to_raw(
+                    MirParser._rechunk_data(
+                        MirParser.convert_raw_to_vis({0: sp_raw}),
+                        [chan_avg],
+                        inplace=True,
+                    )
+                )[0]
+
+        # Finally, return the dict containing the raw data.
+        return data_dict
+
+    def rechunk(self, chan_avg):
+        """
+        Rechunk a MirParser object.
+
+        Spectrally average a Mir dataset. This command attempts to emulate the old
+        "SMARechunker" program within the MirParser object. Users should be aware
+        that running this operation modifies the metadata in such a way that all data
+        loaded will be rechunked in the same manner, until you run the `reset` method.
+
+        Note that this command will only process data from the "normal" spectral
+        windows, and not the pseudo-continuum data (which will remain untouched).
+
+        Parameters
+        ----------
+        chan_avg : int
+            Number of contiguous spectral channels to average over.
+        """
+        # Start of by doing some argument checking.
+        arg_dict = {"chan_avg": chan_avg}
+        for key, value in arg_dict.items():
+            if not (isinstance(value, int) or isinstance(value, np.int_)):
+                raise ValueError("%s must be of type int." % key)
+            elif value < 1:
+                raise ValueError("%s cannot be a number less than one." % key)
+
+        if chan_avg == 1:
+            # This is a no-op, so we can actually bail at this point.
+            return
+
+        if not self._check_data_index():
+            # If the above returns False, then we have a problem, and can't
+            # actually run this operation (have to reload the data).
+            raise ValueError(
+                "Index values do not match data keys. Data will need to be "
+                "reloaded before continuing using `select(reset=True)`."
+            )
+
+        # Eventually, we can make this handling more sophisticated, but for now, just
+        # make it so that we average every window aside from the pseudo continuum
+        # (win #0) by the same value.
+        band_dict = self.codes_data["band"]
+        chanavg_dict = {
```

Comment:
This is clever, borderline too clever, it's not very easy to understand quickly.